crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1101 @@
+
+    for (inst_name, inst_params, inst_qubits) in synth_circuit {
+        score_instruction(&inst_name, &inst_params, &inst_qubits);
+    }
+    1.0 - gate_fidelities.into_iter().product::<f64>()
+}
+
+// This is the outer-most run function. It is meant to be called from Python
+// in `UnitarySynthesis.run()`.
+#[pyfunction]
+#[pyo3(name = "run_default_main_loop")]
+fn py_run_main_loop(
+    py: Python,
+    dag: &mut DAGCircuit,
+    qubit_indices: &Bound<'_, PyList>,
+    min_qubits: usize,
+    target: &Target,
+    coupling_edges: &Bound<'_, PyList>,
+    approximation_degree: Option<f64>,
+    natural_direction: Option<bool>,
+) -> PyResult<DAGCircuit> {
+    let dag_to_circuit = imports::DAG_TO_CIRCUIT.get_bound(py);
+
+    // Run synthesis recursively over control flow ops, mapping qubit indices
+    let node_ids: Vec<NodeIndex> = dag.op_nodes(false).collect();
+    for node in node_ids {
+        let NodeType::Operation(inst) = &dag.dag()[node] else {
+            panic!("DAG node must be an instruction")
+        };
+        if !inst.op.control_flow() {
+            continue;
+        }
+        let OperationRef::Instruction(py_inst) = inst.op.view() else {
+            panic!("Control flow op must be an instruction")
+        };
+        let raw_blocks: Vec<PyResult<Bound<PyAny>>> = py_inst
+            .instruction
+            .getattr(py, "blocks")?
+            .bind(py)
+            .iter()?
+            .collect();
+        let mut new_blocks = Vec::with_capacity(raw_blocks.len());
+        for raw_block in raw_blocks {
+            let new_ids = dag.get_qargs(inst.qubits).iter().map(|qarg| {
+                qubit_indices
+                    .get_item(qarg.0 as usize)
+                    .expect("Unexpected index error in DAG")
+            });
+            let res = py_run_main_loop(
+                py,
+                &mut circuit_to_dag(
+                    py,
+                    QuantumCircuitData::extract_bound(&raw_block?)?,
+                    false,
+                    None,
+                    None,
+                )?,
+                &PyList::new_bound(py, new_ids),
+                min_qubits,
+                target,
+                coupling_edges,
+                approximation_degree,
+                natural_direction,
+            )?;
+            new_blocks.push(dag_to_circuit.call1((res,))?);
+        }
+        let old_node = dag.get_node(py, node)?.clone();
+        let new_node = py_inst
+            .instruction
+            .bind(py)
+            .call_method1("replace_blocks", (new_blocks,))?;
+        dag.py_substitute_node(old_node.bind(py), &new_node, true, false)?;
+    }
+    let mut out_dag = dag.copy_empty_like(py, "alike")?;
+
+    // Iterate over dag nodes and determine unitary synthesis approach
+    for node in dag.topological_op_nodes()? {
+        let NodeType::Operation(packed_instr) = &dag.dag()[node] else {
+            panic!("DAG node must be an instruction")
+        };
```

Comment:
You can use `unwrap_operation` here as well:
```suggestion
        let packed_instr = dag.dag()[node].unwrap_operation();
```